stackrox/acs-fleet-manager
# ROX-16506: Add Fleet-Manager OperatorImage configuration

## Description

This PR adds the fleet-manager logic to use fleet-manager to manage Central and Operator versions.

**Workflow:**

The workflow implementation is based declaratively in a sense that Fleet-Manager defines the desired state, but does not take care of status reporting. Status reporting is implemented via alerting later.

 1. CentralRequest is created and fetched default image from the default version table (table to be renamed by @johannes94 ), see [here](https://github.com/stackrox/acs-fleet-manager/pull/1028/files#diff-080490e019286b765c01a138eab3a47c8dd81e8c5c942fc8545be9d8fa8ae4e9R252-R257)
 2. `operator_image` field is set on creation
 3. Fleetshard-Sync sees new operator image in its runtime
 4. Fleetshard-Sync creates for each operator-image a new deployment

**Changes different from design document:**

 - `previous_*` fields were removed
 - `actual_*` fields were removed
 - `operator_image` field was added

_Status reporting_
Status reporting needs separate implementation and design to have a sustainable solution to query states in fleet-manager. My biggest concern is that structured status reporting creates an overhead in adding new statuses to fleet-manager,i .e. by adding new database fields and reconcilation logic.

_Previous Image_
To rollback an instance we need to store the previous revision of a Central. Instead of as initially saving only the `previous_image` the whole CentralRequest object can be revisioned.  (Ticket: https://issues.redhat.com/browse/ROX-8987)
This may need a small implementation discussion.

_End to end tests_

End-to-end, integration and unit-tests are added in a later PR. Currently all features are hidden behind a feature flag.

**Removals:**

 - Some operator version related code was removed which was not used anymore because of the changes to the fields.
 - Follow up removal: acceptance manager
 - Follow up removal: operator & central version legacy logic

**DoD and follow-ups:**

 - [x] Set default version when instance is created
 - [x] Add image fields to Central table
 - [x] Fleetshard should see new fields and install operator based on these
 - [x] Migrate existing Centrals (ticket: https://issues.redhat.com/browse/ROX-17759)
 - [x] Fetch operator images from Fleet-Manager
 - [x] Keep history of previously applied operator image to an instance for possible rollbacks (https://issues.redhat.com/browse/ROX-8987)
 - [x] use correct label selector to support multiple operator instances (ticket: https://issues.redhat.com/browse/ROX-17718)

## TODO:

## Checklist (Definition of Done)


## Test manual

 - Install fleet-manager locally via `./scripts/install_operator_canary.sh`
 - Create stackrox-operator namespace
 - Add `quay-ips` secret manually to stackrox-operator namespace
 - Create Central via `./scripts/create-central.sh`
 - Run `kubectl get pods -n stackrox-operator` to see new installed operators
